packages/next-server/lib/router/router.js

```diff
@@ -154,39 +154,42 @@ export default class Router {
 
     this.abortComponentLoad(as)
 
+    const { pathname: asPathname, query: asQuery } = parse(as, true)
+    const { pathname, query } = parse(url, true)
+    const route = toRoute(pathname)
+
+    // Reuse existing routeInfo when possible.
+    let routeInfo = this.components[route] || null
+
     // If the url change is only related to a hash change
     // We should not proceed. We should only change the state.
     if (this.onlyAHashChange(as)) {
       Router.events.emit('hashChangeStart', as)
       this.changeState(method, url, as)
       this.scrollToHash(as)
+
+      // record hash change
+      const hash = window.location.hash.substring(1)
+      this.set(route, pathname, query, as, { ...routeInfo, hash })
+
       Router.events.emit('hashChangeComplete', as)
       return true
     }
 
-    const { pathname: asPathname, query: asQuery } = parse(as, true)
-    const { pathname, query } = parse(url, true)
-
-    // If asked to change the current URL we should reload the current page
-    // (not location.reload() but reload getInitialProps and other Next.js stuffs)
+    // If asked to change the current URL, we should reload the current page component.
     // We also need to set the method = replaceState always
     // as this should not go into the history (That's how browsers work)
     if (!this.urlIsNew(asPathname, asQuery)) {
       method = 'replaceState'
     }
 
-    const route = toRoute(pathname)
-    const { shallow = false } = options
-    let routeInfo = null
-
     Router.events.emit('routeChangeStart', as)
 
-    // If shallow === false and other conditions met, we reuse the
-    // existing routeInfo for this route.
-    // Because of this, getInitialProps would not run.
-    if (shallow && this.isShallowRoutingPossible(route)) {
-      routeInfo = this.components[route]
-    } else {
+    const { shallow = false } = options
+
```

Comment:
Kept this block down here to avoid accidentally calling getInitialProps for a hash change (although it should be noted that we won't call `this.scrollToHash(as)` when we're navigating between pages with an attached hash. 

There's an open issue for that, though.